xmake/rules/c++/modules/modules_support/msvc/builder.lua

```diff
@@ -0,0 +1,335 @@
+--
+-- Copyright (C) 2015-present, TBOOX Open Source Group.
+--
+-- @author      ruki, Arthapz
+-- @file        msvc/builder.lua
+--
+
+-- imports
+import("core.base.json")
+import("core.base.option")
+import("core.base.semver")
+import("utils.progress")
+import("private.action.build.object", {alias = "objectbuilder"})
+import("core.tool.compiler")
+import("core.project.config")
+import("core.project.depend")
+import("private.tools.vstool")
+import("compiler_support")
+import(".builder", {inherit = true})
+
+-- get flags for building a module
+function _make_modulebuildflags(target, provide, bmifile, sourcefile, objectfile, opt)
+
+    -- get flags
+    local ifcoutputflag = compiler_support.get_ifcoutputflag(target)
+    local ifconlyflag = compiler_support.get_ifconlyflag(target)
+    local interfaceflag = compiler_support.get_interfaceflag(target)
+    local internalpartitionflag = compiler_support.get_internalpartitionflag(target)
+
+    local objectfile_or_ifconlyflag = (opt.external and ifconlyflag) and ifconlyflag or "-Fo" .. objectfile
+
+    local flags
+    if provide then -- named module
+        flags = {"-TP", ifcoutputflag, bmifile, provide.interface and interfaceflag or internalpartitionflag, objectfile_or_ifconlyflag, "-c", sourcefile}
+    else
+        flags = {"-TP", objectfile_or_ifconlyflag, "-c", sourcefile}
+    end
+
+    return flags
+end
+
+-- get flags for building a headerunit
+function _make_headerunitflags(target, headerunit, bmifile, opt)
+
+    -- get flags
+    local exportheaderflag = compiler_support.get_exportheaderflag(target)
+    local headernameflag = compiler_support.get_headernameflag(target)
+    local ifcoutputflag = compiler_support.get_ifcoutputflag(target)
+    assert(headernameflag and exportheaderflag, "compiler(msvc): does not support c++ header units!")
+
+    local local_directory = (headerunit.type == ":quote") and {"-I" .. path.directory(headerunit.path)} or {}
+
+    local flags = table.join(local_directory, {"-TP", 
+                                               exportheaderflag,
+                                               headernameflag .. headerunit.type,
+                                               ifcoutputflag,
+                                               bmifile,
+                                               headerunit.type == ":angle" and headerunit.name or headerunit.path})
+    return flags
+end
+
+-- do compile
+function _compile(target, flags, sourcefile)
+
+    local dryrun = option.get("dry-run")
+    local compinst = target:compiler("cxx")
+    local compflags = compinst:compflags({sourcefile = sourcefile, target = target})
+    local msvc = target:toolchain("msvc")
+    local flags = table.join(compflags or {}, flags)
+
+    -- trace
+    if option.get("verbose") then
+        print(os.args(table.join(compinst:program(), flags)))
+    end
+
+    if not dryrun then
+        -- do compile
+        os.vrunv(compinst:program(), flags, {envs = msvc:runenvs()})
```

Comment:
why not use compinst:compile

Reply:
https://github.com/xmake-io/xmake/pull/4321#discussion_r1472792701